engine/src/flutter/shell/platform/android/test/io/flutter/embedding/engine/loader/FlutterLoaderTest.java

```diff
@@ -399,4 +405,151 @@ public void itSetsShaderInitModeFromMetaData() {
+        internalStorageDirAsPathObj.resolve(
+            Paths.get("some", "directory", "..", "..", "library.so"));
+    Path pathWithIndirectInternalStoragePath3 =
+        internalStorageDirAsPathObj.resolve(Paths.get("some", "directory", "..", "library.so"));
+
+    Path[] pathsToTest = {
+      pathWithDirectInternalStoragePath,
+      pathWithNestedInternalStoragePath,
+      pathWithIndirectInternalStoragePath1,
+      pathWithIndirectInternalStoragePath2,
+      pathWithIndirectInternalStoragePath3
+    };
+    String aotSharedNameArgPrefix = "--aot-shared-library-name=";
+
+    for (Path testPath : pathsToTest) {
+      String path = testPath.toString();
+      String aotSharedLibraryNameArg = aotSharedNameArgPrefix + path;
+      String[] args = {aotSharedLibraryNameArg};
+      flutterLoader.ensureInitializationComplete(ctx, args);
+      shadowOf(getMainLooper()).idle();
+
+      ArgumentCaptor<String[]> shellArgsCaptor = ArgumentCaptor.forClass(String[].class);
+      verify(mockFlutterJNI)
+          .init(
+              eq(ctx),
+              shellArgsCaptor.capture(),
+              anyString(),
+              anyString(),
+              anyString(),
+              anyLong(),
+              anyInt());
+
+      List<String> actualArgs = Arrays.asList(shellArgsCaptor.getValue());
+
+      // This check works because the tests run in debug mode. If run in release (or JIT release)
+      // mode, actualArgs would contain the default arguments for AOT shared library name on top
+      // of aotSharedLibraryNameArg.
+      String canonicalTestPath = testPath.toFile().getCanonicalPath();
+      String canonicalAotSharedLibraryNameArg = aotSharedNameArgPrefix + canonicalTestPath;
+      assertTrue(
+          "Args sent to FlutterJni.init incorrectly did not include path " + path,
+          actualArgs.contains(canonicalAotSharedLibraryNameArg));
+
+      // Reset FlutterLoader and mockFlutterJNI to make more calls to
+      // FlutterLoader.ensureInitialized and mockFlutterJNI.init for testing.
+      flutterLoader.initialized = false;
+      clearInvocations(mockFlutterJNI);
+    }
+  }
+
+  @Test
+  public void itDoesNotSetAotSharedLibraryNameIfPathOutsideInternalStorage() throws IOException {
+    FlutterJNI mockFlutterJNI = mock(FlutterJNI.class);
+    FlutterLoader flutterLoader = spy(new FlutterLoader(mockFlutterJNI));
+    Context mockApplicationContext = mock(Context.class);
+    File internalStorageDir = ctx.getFilesDir();
+    Path internalStorageDirAsPathObj = internalStorageDir.toPath();
+    String internalStorageDirPath = internalStorageDir.getCanonicalPath();
+
+    ctx.getApplicationInfo().nativeLibraryDir =
+        Paths.get("some", "path", "doesnt", "matter").toString();
+    assertFalse(flutterLoader.initialized());
+    flutterLoader.startInitialization(ctx);
+
+    // Test paths for library living outside internal storage.
+    Path pathThatIsCompletelyUnrelated = Paths.get("please", "dont", "fail");
```

Comment:
I think this path should fail. Consider renaming the path 
```suggestion
    Path pathThatIsCompletelyUnrelated = Paths.get("please", "do", "fail");
```